gateway/ln-gateway/src/gatewayd/actor.rs

```diff
@@ -147,7 +154,36 @@ impl GatewayActor {
 
                     let amount_msat = Amount::from_msats(outgoing_amount_msat);
 
-                    match actor.buy_preimage_internal(&hash, &amount_msat).await {
+                    let (outpoint, contract_id) =
+                        match actor.buy_preimage_internal(&hash, &amount_msat).await {
+                            Ok((outpoint, contract_id)) => (outpoint, contract_id),
+                            Err(e) => {
+                                error!("Failed to buy preimage: {:?}", e);
+                                // Note: this specific complete htlc requires no futher action.
+                                // If we fail to send the complete htlc message, or get an error
+                                // result, lightning node will still
+                                // cancel HTCL after expiry period lapses.
+                                // Result can be safely ignored.
+                                // TODO: make sure this succeeded?
+                                let _ = lnrpc_copy
+                                    .complete_htlc(CompleteHtlcsRequest {
+                                        intercepted_htlc_id,
+                                        action: Some(Action::Cancel(Cancel {
+                                            reason: e.to_string(),
+                                        })),
+                                    })
```

Comment:
In the future we should add test coverage to trigger these failure case and test the it behaves as expected. But I don't think this should block this PR because it does pass our existing test suite which means we aren't testing this sort of incoming HTLC failures.

Also, we shouldn't cheat with `let _ = <result>` in the future.

Reply:
retry of the rpc call could come in handy

Reply:
> retry of the rpc call could come in handy. We should retry sending the rpc call to cancel htlc

:+1: 

the `utils::retry` function might be useful here, right?